src/main/java/com/spotify/futures/FailFast.java

```diff
@@ -0,0 +1,60 @@
+package com.spotify.futures;
+
+/**
+ * Interface that defines fail fast behavior when combining several completion stages in a
+ * completable future.
+ *
+ * @author Jose Alavez
+ * @since 0.3.3
+ */
+public interface FailFast {
+
+  /**
+   * Instance used to avoid fail fast behavior when combining futures.
+   */
+  FailFast NO_FAILFAST = new FailFast() {
+    @Override
+    public boolean shouldFailFast(Throwable throwable, ExecutionMetadata executionMetadata) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Throwable withThrowable(Throwable origin) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean cancelAll() {
+      throw new UnsupportedOperationException();
+    }
+  };
+
+  /**
+   * Evaluates if the {@code throwable} should lead to a fast failure when combining completion
+   * stages together. A fast failure will not wait for all the combined stages to complete normally
+   * or exceptionally.
+   *
+   * @param throwable         A {@link Throwable} returned by a exceptionally completed stage.
+   * @param executionMetadata A {@link ExecutionMetadata} that provides data related to current
+   *                          execution.
+   * @return {@code true} if the combined completable future should fail fast, {@code false} if
+   * otherwise.
+   */
+  boolean shouldFailFast(Throwable throwable, ExecutionMetadata executionMetadata);
+
+  /**
+   * Specifies a {@link Throwable} instance to return when a combined completable future fail fast.
+   *
+   * @param origin A {@link Throwable} that originated the fail fast.
+   * @return A {@link Throwable} to used to complete exceptionally a combined completable future.
+   */
+  Throwable withThrowable(Throwable origin);
+
+  /**
+   * Defines if all the incomplete stages should be cancelled when failing fast.
+   *
+   * @return {@code true} if all the incomplete stages should be cancelled when failing fast, {@code
+   * false} if otherwise.
+   */
+  boolean cancelAll();
```

Comment:
+1 to look into this refactoring. Or actually, I'm wondering if we can get away without this abstraction at all and just have a single implementation that always cancels and fails on every exception.

Reply:
First and foremost, thank you for your valuable opinion!

As explained here https://github.com/spotify/completable-futures/pull/68/files#r434177315, I think I was a little shortsighted when I created this interface. My end goal with it's creation was to allow extension of behavior on when and how to fail fast, but it seems I missed the `how` within this implementation.

In the project we use this library (`completable-futures`) we are most likely wanting to fail fast if an exception happened, but also considering that this exception "could be forgiven" if the majority of tasks succeeded. In that aspect, I think @dflemstr could provide greater control.

Reply:
@mbruggmann I agree this could be solved with a specific implementation rather than an abstraction to inject behavior. Sadly, I don't think this would be beneficial in my user case.

We use `completable-futures` as an alternative to the creation of `batch rest get` endpoints in the services we consume. If we compare the behavior of a batch endpoint, `not found` items are usually excluded from a resulting collection, rather than causing 5xx. 

We are likely needing to be able to specify which exceptions could cause the overall result to complete exceptionally (as fast as possible), while in other cases, be forgiven and prefer the majority of completed results while skipping a few (due to for example `not found`).

Reply:
Right, I see what you mean. Given that `allAsList` and `joinList` will always fail if any of their inputs fail (unlike `successfulAsList`), I think we'd need to find a different solution for this. Maybe that can happen before the `joinList`?
```java
collection.stream()
    .map(this::someAsyncFunction)
    .map(stage -> stage.handle(...)) // <-- fallback to a successful stage in certain cases
    .collect(CompletableFutures.joinList())
    .thenApply(this::consumeList)
```

Reminds me a bit of [Futures.catching(...) from Guava](https://github.com/google/guava/blob/master/guava/src/com/google/common/util/concurrent/Futures.java#L294). Maybe we could add something similar? Would suggest a separate PR for that though.

Reply:
Indeed using `handle` would allow to default (arguably to null as you cannot simply generate empty in https://github.com/spotify/completable-futures/blob/32826e43ad3eb94131513efd59691bf5065c332c/src/main/java/com/spotify/futures/CompletableFutures.java#L110) exceptions that are known beforehand to be admissible to fail. 

A more complicated scenario would be a Timeout (408), there the decision of what do could be based on the current state of all the other executions (I would still argue that failing exceptionally due to 1 out of hundreds of tasks, would be not desirable).

But while giving this a second round of thought, maybe I'm asking too much :). A fail fast behavior on *any exception* and that always *cancels* remaining tasks would still be beneficial to us. Something like `joinListFailingFast()` or `joinList(boolean failFast)`.